lib/Runtime/Library/JavascriptLibrary.cpp

```diff
@@ -5276,16 +5310,58 @@ namespace Js
 
     void JavascriptLibrary::SetCrossSiteForLockedNonBuiltInFunctionType(JavascriptFunction * function)
     {
+        FunctionProxy * functionProxy = function->GetFunctionProxy();
         DynamicTypeHandler *typeHandler = function->GetTypeHandler();
-        if (typeHandler->IsPathTypeHandler())
+        if (typeHandler->IsDeferredTypeHandler())
         {
-            PathTypeHandlerBase::FromTypeHandler(typeHandler)->ConvertToNonShareableTypeHandler(function);
+            if (functionProxy && functionProxy->GetCrossSiteDeferredFunctionType())
+            {
+                function->ReplaceType(functionProxy->GetCrossSiteDeferredFunctionType());
+            }
+            else
+            {
+                function->ChangeType();
+                function->SetEntryPoint(scriptContext->CurrentCrossSiteThunk);
+                if (functionProxy && !PHASE_OFF1(ShareCrossSiteFuncTypesPhase))
+                {
+                    function->ShareType();
+                    functionProxy->SetCrossSiteDeferredFunctionType(UnsafeVarTo<ScriptFunction>(function)->GetScriptFunctionType());
+                }
+            }
         }
-        else
+        else 
         {
-            function->ChangeType();
+            if (functionProxy && functionProxy->GetCrossSiteUndeferredFunctionType())
+            {
+                function->ReplaceType(functionProxy->GetCrossSiteUndeferredFunctionType());
+            }
+            else
+            {
+                if (typeHandler->IsPathTypeHandler())
+                {
+                    if (!PHASE_OFF1(ShareCrossSiteFuncTypesPhase))
+                    {
+                        DynamicType *type = function->DuplicateType();
```

Comment:
>DuplicateType [](start = 54, length = 13)

Are we calling DuplicateType just to get a newly allocated type? I ask because ScriptFunction::DuplicateType will register the new type in the `functionObjectTypeList` on the function proxy. 

Now, I don't know what the correct thing to do is here, I'm just curious as to what the intention of calling DuplicateType is.

Reply:
Yes, we want a new type, and I think having it in the function object type list is necessary. Note that ChangeType calls DuplicateType. The reason for calling DuplicateType directly here is that we don't necessarily want to install this type on the object; we may only want it to serve as the root of a new path.

Reply:
Alrighty. Thanks.

---
In reply to: [231686357](https://github.com/Microsoft/ChakraCore/pull/5794#discussion_r231686357) [](ancestors = 231686357)